src/mds/SessionMap.cc

```diff
@@ -449,3 +656,41 @@ void Session::decode(bufferlist::iterator &p)
   _update_human_name();
 }
 
+/**
+ * Mark a session as requiring write by the current value of `version`.
+ */
+void SessionMap::mark_dirty(const Session *s)
+{
+  dout(20) << __func__ << " s=" << s << " name=" << s->info.inst.name << dendl;
+  dirty_sessions.insert(s->info.inst.name);
+}
+
+/**
+ * Start a new version: this will make the 
+ */
+void SessionMap::inc_version()
+{
+  dout(20) << __func__ << " v " << version << " -> " << version + 1
+    << " dirty=" << dirty_sessions.size() << "/"
+    << g_conf->mds_sessionmap_keys_per_op / 2 << dendl;
+
+  if (dirty_sessions.size() >= g_conf->mds_sessionmap_keys_per_op / 2) {
+    // If there are >= allowance/2 sessions dirty, then we have to
+    // do a save now, because otherwise the sessions written in this
+    // newly incremented version could leave us with more than `allowance`
+    // dirty sessions, and thereby an oversized write in the next save()
+    save(new C_MDSInternalNoop, version);
```

Comment:
I toyed with the idea of tracking extra state like this.  It would mean mean adding a mark_projected equivalent of mark dirty, and tracking a history of which versions a session had been projected into -- feels a bit overkill to me, as we don't have a functional need to track per-version sessions, just to barrier the writes of sessions against the global version (if that sentence made any sense at all)